### ISSUE TITLE
Avoid empty Markdown page get generated

### ISSUE THREAD

**Issue by BLumia** [doxygen/doxygen#9437]
**Describe the bug**
An empty page will be generated if there is a markdown file with empty content.

**Expected behavior**
Don't generate the empty related pages, or offer an option to avoid generating related pages with empty content. 

**To Reproduce**
[notdtkcore.zip](https://github.com/doxygen/doxygen/files/9060705/notdtkcore.zip) (edit: `-DDOXYGEN_GENERATE_HTML="YES"` needs to be pass to `cmake` since the default value is "NO" from this example)

Here is a simple test project that can be used to reproduce this issue. Currently it tries to generate Chinese documentation (you can change to English and you will get similar result). I'll explain the thing I want to achieve in **Additional context** section.

**Version**

```
$ doxygen --version
1.9.3
$ lsb_release -a
LSB Version:    n/a
Distributor ID: Arch
Description:    Arch Linux
Release:        rolling
Codename:       n/a
```

**Additional context**
The main goal is we want to provide both English and Chinese version of documentation for our project, and we want to avoid writing a lot of `@~langId` in our source file. As a solution, we decided to write documentation outside our source file, so one `.dox`/`.md` file only contain documentation for a single language, so we don't need to write a lot of `@~langId` in our documentation files, only one will be enough.

Then we found, for `.dox` file it's okay since it's documentation for APIs, but for `.md` file it isn't. If we generate English documentation by set the OUTPUT_LANGUAGE (via `DOXYGEN_OUTPUT_LANGUAGE`), then in the demo project, an empty `index.zh_CN` page will be generated and listed under Related Pages tab.

As a workaround solution, we can use FILE_PATTERNS (via `DOXYGEN_FILE_PATTERNS`) and filter `.zh_CN.md` when we generate Chinese documentation or set to `.en_US.md` when we generate English documentation, so I file this bug and would like to ask, could we have an option to let doxygen avoid generate empty pages, or do we have a better solution for this purpose?

Thanks!

**Comment by BLumia** [comment]
By the way, I also notice [the first level 1 header will be used as the title of the page and a header id attribution can be set](https://doxygen.nl/manual/markdown.html#md_page_header). This can be used to avoid generating duplicated pages if we have 2 translation markdown documents for the same page, which can also be a good solution for this issue IMO, but this only works if there are nothing above the level 1 header, so something like

``` markdown
@~english
Test Page                         {#testpage}
============
```

or

``` markdown
@~english
# Test Page #          {#testpage}
```

or

``` markdown
@~english
@page testpage Test Page
```

won't work. If this is working, then I guess this issue can also be considered as resolved.

**Comment by albert-github** [comment]
I think we have here 2 separate and independent issues.
1. problem with remaining empty files (original description https://github.com/doxygen/doxygen/issues/9437#issue-1296869866)
2. problem with `@~...` blocking the automatic page name generation (the first comment https://github.com/doxygen/doxygen/issues/9437#issuecomment-1177142471)

- Please create a new issue for the second problem (example can be the same).

Regarding the first problem, this is a valid issue this is a valid issue, though in your case and as you use cmake to generate your `Doxyfile` you can make the CMakeLists.txt file also so intelligent to exclude the files that you don't need by means of e.g. `EXCLUDE_PATTERNS=*/*en_US*` when generating the Chinese language version. This would probably also overcome the problem of the need of the `@~...` at the beginning of the file.


Personally I think the decision "As a solution, we decided to write documentation outside our source file" is not a good decision as now one has to change 3 files and the files could easily diverge.

**Comment by albert-github** [comment]
Regarding the second problem see also the remark at https://github.com/doxygen/doxygen/pull/6909#issuecomment-1177453340

**Comment by BLumia** [comment]
> This would probably also overcome the problem of the need of the @~... at the beginning of the file.

Yes, so currently I'm using the *.zh_CN.md pattern to include all Chinese documentation and use *.en_US.md for all English documentation, works in a same way :)

> Personally I think the decision "As a solution, we decided to write documentation outside our source file" is not a good decision as now one has to change 3 files and the files could easily diverge.

I agree it's diverge but there are also something we want to avoided. Documenting directly in code will results non-ASCII characters in the source file, which might cause compile issue in some specific compiler. We do not really want to add a lots of `\~english` commands in our source file (actually if there is a global one that can be used in the source file insead of per-comment-block then we may keep the English documentation in the source file), so we also decided to move the English documentation outside the source files.

I also tested the [`coverxygen`](https://github.com/psycofdj/coverxygen) tool and hppe adding a CI could ensure newly added APIs got documentation covered, hopefully it could reduce some issues caused by the diverge.

> Regarding the second problem see also the remark at https://github.com/doxygen/doxygen/pull/6909#issuecomment-1177453340

Good to know, do I still need to create another issue thread for the 2nd issue?

**Comment by BLumia** [comment]
> It is better to have a separate issue for it as both problems are unrelated.

I though it might be consider duplicated with https://github.com/doxygen/doxygen/pull/6909#issuecomment-1177453340 😂 

Anyway for the 2nd issue, see https://github.com/doxygen/doxygen/issues/9438.

**Comment by albert-github** [comment]
Yes please do create a standalone, minimal example and also mention the doxygen version used (looks like to be 1.12.0).

Small note: I would not use the Table of contents as shown in your example but either use a TOC as you already do in https://doc.riot-os.org/index.html or create a TOC with md references (`[...](...)`) as doxygen can create automatic links in that case (see also `MARKDOWN_ID_STYLE=github`).

The repository you use / refer to is I think https://github.com/RIOT-OS/RIOT, does this already contain the mentioned changes?

**Comment by crasbe** [comment]
Attached you find a minimal example, I used the doxygen 1.14.0 which is still installed from the last issue.

[Doxyfile.txt](https://github.com/user-attachments/files/19088288/Doxyfile.txt)
[doc.md](https://github.com/user-attachments/files/19088350/doc.md)

The Doxyfile is a standard Doxyfile, the only changes are these two options:
```
GENERATE_LATEX = NO
FILE_PATTERNS          = *.txt \
                         doc.md \
                         *.h \
                         *.h.in \
                         *.hpp
```

As you can see, the empty `doc` page is still generated...
![Image](https://github.com/user-attachments/assets/91571350-5955-4ae4-ab46-b34b7853dcd1)

...and the content of the file is put into the group that was defined.
![Image](https://github.com/user-attachments/assets/3fe3104a-4ea3-48fa-8cf8-2f285e511c64)

<hr/>

Yes, the remark about the table of contents is valid. I'll add it to our Todo list. Thank you!

<hr/>

Yes, we applied the change with https://github.com/RIOT-OS/RIOT/pull/21255 after the discussion in https://github.com/RIOT-OS/RIOT/issues/21220

We can revert the change, so it is not super time critical, but in general a solution in Doxygen would probably be most desirable (as others had the same issue previously).

**Comment by albert-github** [comment]
In commit:
```
Commit: fd23262193b9e528d44b0067bda0341bb3d0d11c [fd23262]

Date: Friday, October 25, 2024 4:13:18 PM

issue #11179 Using README.md as directory description.
```

a.o. the `@dir` command has been added for markdown files as they can be explicit pages.
I think the same reasoning for `@dir` is valid for `@defgroup` (and `@addtogroup`).

Maybe there are even more commands that should be handled similar.

**Comment by albert-github** [comment]
Regarding the `defgroup` problem I've just pushed a proposed patch, pull request #11478

**Comment by crasbe** [comment]
I can confirm that the fix works as intended. Both in the minimal example as well as in the RIOT documentation, no empty `doc` page is generated anymore. Thank you very much @albert-github :)

![Image](https://github.com/user-attachments/assets/40fc0487-4538-4b43-a1b8-2e023d79febb)
![Image](https://github.com/user-attachments/assets/b83073e3-f891-40cc-bc96-567cd14a132a)
![Image](https://github.com/user-attachments/assets/86debe30-8406-401f-b52c-435f1c6746b0)

I wonder if it would be good to add this behavior to the documentation:

```diff

```

**Comment by albert-github** [comment]
Would be nice to have a documentation PR for it though I would not list the commands explicitly (hard to maintain as when adding commands this has to be added at multiple places, code and documentation) but use something like "dedicated command (a.o. `\defgroup`, `\dir`)"

**Comment by albert-github** [comment]
I didn't expect this behavior, but I think there are reasons why it is not working / intended behavior.

I don't think you chose a bad example but just a good one so it shows a potential problem.

2 small notes:
- The `\brief` description is shown twice probably due to the default setting `REPEAT_BRIEF=YES` so best would be to either remove the `\brief` here or set `REPEAT_BRIEF=NO`.
- You now use something like 
  ```
  1. [Overview](#esp32_overview)
  ```
  and in the code
  ```
  # Overview {#esp32_overview}
  ```
  You could consider to use
  - `-#` instead of `1.` etc
  - setting `MARKDOWN_ID_STYLE=github` and use
    ```
    [Overview](#overview)
    ```
    and
    ```
    # Overview
    ```

  Disclaimer I didn't test the second advise so it may contain some typos

**Comment by albert-github** [comment]
Solved original issue as well in the proposed pull request #11478

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by albert-github** [comment]
In `subgroup.md` you use:
```
@defgroup    basegroup_subgroup Subgroup of Basegroup
@addtogroup  basegroup
@brief       This is a subgroup of the basegroup.
```

this should be:
```
@defgroup    basegroup_subgroup Subgroup of Basegroup
@ingroup  basegroup
@brief       This is a subgroup of the basegroup.
```

**Comment by crasbe** [comment]
But `@ingroup` will create a reference to the Subgroup in the Basegroup page, whereas `@addtogroup` will include the whole Subgroup in the Basegroup page.

In case the latter behavior is the desired one (for example for splitting a large documentation file up into multiple files that should be rendered into one page), it would leave the empty page.

**Comment by albert-github** [comment]
I see.

In that case the `@defgroup    basegroup_subgroup Subgroup of Basegroup` has no meaning and and it is not necessary.
The page is not really empty as you explicitly defined a group, so I think the implementation is correct.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.14.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
